cmd/kubelet/app/server.go

```diff
@@ -476,9 +476,17 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
 		}
 
 		kubeDeps.Cloud = cloud
-		kubeDeps.KubeClient = kubeClient
-		kubeDeps.ExternalKubeClient = externalKubeClient
-		kubeDeps.EventClient = eventClient
+		// we can not set struct pointer directly to interface when struct pointer is nil
+		// if we did that the expression (interface == nil) will never equal
+		if kubeClient != nil {
```

Comment:
The default value of these interface in function `UnsecuredKubeletDeps` is nil:
```
return &kubelet.KubeletDeps{
		Auth:               nil, // default does not enforce auth[nz]
		CAdvisorInterface:  nil, // cadvisor.New launches background processes (bg http.ListenAndServe, and some bg cleaners), not set here
		Cloud:              nil, // cloud provider might start background processes
		ContainerManager:   nil,
		DockerClient:       dockerClient,
		KubeClient:         nil,
		ExternalKubeClient: nil,
		EventClient:        nil,
		Mounter:            mounter,
		NetworkPlugins:     ProbeNetworkPlugins(s.NetworkPluginDir, s.CNIConfDir, s.CNIBinDir),
		OOMAdjuster:        oom.NewOOMAdjuster(),
		OSInterface:        kubecontainer.RealOS{},
		Writer:             writer,
		VolumePlugins:      ProbeVolumePlugins(s.VolumePluginDir),
		TLSOptions:         tlsOptions,
	}, nil

```

Point of this change is to make sure the default value(nil) of `kubeDeps.KubeClient` not changed by setting `kubeDeps.KubeClient = kubeClient`  when kubeClient = nil.

and if we directly use `kubeDeps.KubeClient = kubeClient` not checking kubeClient value,
that will cause panic in function `makeEventRecorder` ,the line code is:
```
if kubeDeps.EventClient != nil {
		glog.V(4).Infof("Sending events to api server.")
		eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeDeps.EventClient.Events("")})
	} else {
		glog.Warning("No api server defined - no events will be sent to API server.")
	}
```
the condition of` kubeDeps.EventClient != nil `will always true, even if the previous `eventClient = nil`